File: adapters/football_edge/pip_client.php
<?php
/**
 * Fail-soft PIP client for Domeneshop/PHP-compatible FEA deployments.
 *
 * This client only performs read-only GET calls to PIP. It never writes, never
 * executes bookmaker actions and never enables real-money or auto-betting logic.
 */

function pip_env_bool(string $name, bool $default = false): bool {
    $value = getenv($name);
    if ($value === false || $value === '') {
        return $default;
    }
    return in_array(strtolower(trim($value)), ['1', 'true', 'yes', 'on'], true);
}

function football_edge_get_pip_probability(int $fixtureId, string $market = '1X2'): array {
    $enabled = pip_env_bool('PIP_INTEGRATION_ENABLED', false);
    $failsafe = pip_env_bool('PIP_FAILSAFE_MODE', true);
    $useMock = pip_env_bool('PIP_USE_MOCK', false);
    $baseUrl = rtrim((string) getenv('PIP_BASE_URL'), '/');
    $apiKey = (string) getenv('PIP_API_KEY');
    $timeout = (int) (getenv('PIP_TIMEOUT_SECONDS') ?: 5);
    if ($timeout < 1) {
        $timeout = 5;
    }

    if (!$enabled) {
        return [
            'enabled' => false,
            'available' => false,
            'fallback_used' => true,
            'data' => null,
            'error_code' => 'PIP_DISABLED',
            'error_message' => 'PIP integration is disabled by configuration.',
            'latency_ms' => null,
        ];
    }

    if ($useMock) {
        $mockPath = dirname(__DIR__, 2) . '/examples/mock_pip_probability_response.json';
        if (is_readable($mockPath)) {
            $payload = json_decode(file_get_contents($mockPath), true);
            $payload['fixture_id'] = $fixtureId;
            $payload['market'] = $market;
            return [
                'enabled' => true,
                'available' => true,
                'fallback_used' => false,
                'data' => $payload,
                'error_code' => null,
                'error_message' => null,
                'latency_ms' => 0,
            ];
        }
        return pip_failsoft_result('PIP_MOCK_LOAD_FAILED', 'Mock PIP response could not be loaded.');
    }

    if ($baseUrl === '') {
        return pip_failsoft_result('PIP_BASE_URL_MISSING', 'PIP_BASE_URL is not configured.');
    }

    $url = $baseUrl . '/api/v1/probability/football/fixture/' . rawurlencode((string) $fixtureId) . '?market=' . rawurlencode($market);
    $start = microtime(true);

    try {
        $headers = ["Accept: application/json"];
        if ($apiKey !== '') {
            $headers[] = "X-PIP-API-Key: " . $apiKey;
        }

        if (function_exists('curl_init')) {
            $ch = curl_init($url);
            curl_setopt_array($ch, [
                CURLOPT_RETURNTRANSFER => true,
                CURLOPT_TIMEOUT => $timeout,
                CURLOPT_HTTPHEADER => $headers,
            ]);
            $body = curl_exec($ch);
            $httpCode = (int) curl_getinfo($ch, CURLINFO_HTTP_CODE);
            $curlError = curl_error($ch);
            curl_close($ch);
            if ($body === false || $httpCode !== 200) {
                return pip_failsoft_result('PIP_HTTP_ERROR', 'PIP HTTP request failed: ' . ($curlError ?: $httpCode));
            }
        } else {
            $context = stream_context_create([
                'http' => [
                    'method' => 'GET',
                    'header' => implode("\r\n", $headers),
                    'timeout' => $timeout,
                ],
            ]);
            $body = @file_get_contents($url, false, $context);
            if ($body === false) {
                return pip_failsoft_result('PIP_HTTP_ERROR', 'PIP HTTP request failed with stream fallback.');
            }
        }

        $payload = json_decode($body, true);
        if (!is_array($payload)) {
            return pip_failsoft_result('PIP_INVALID_JSON', 'PIP returned invalid JSON.');
        }
        if (($payload['execution_allowed'] ?? null) !== false) {
            return pip_failsoft_result('PIP_CONTRACT_EXECUTION_FLAG', 'PIP violated execution_allowed=false contract.');
        }

        return [
            'enabled' => true,
            'available' => true,
            'fallback_used' => false,
            'data' => $payload,
            'error_code' => null,
            'error_message' => null,
            'latency_ms' => (int) round((microtime(true) - $start) * 1000),
        ];
    } catch (Throwable $e) {
        if (!$failsafe) {
            throw $e;
        }
        return pip_failsoft_result('PIP_CLIENT_EXCEPTION', $e->getMessage());
    }
}

function pip_failsoft_result(string $code, string $message): array {
    return [
        'enabled' => true,
        'available' => false,
        'fallback_used' => true,
        'data' => null,
        'error_code' => $code,
        'error_message' => $message,
        'latency_ms' => null,
    ];
}
?>

File: adapters/football_edge/pip_client.py
"""Fail-soft PIP client for Football Edge Agent.

Design intent:
- disabled by default
- short timeout
- no write endpoints
- no execution capability
- never throws into the FEA recommendation path when failsafe mode is enabled
"""
from __future__ import annotations

import json
import os
import time
import urllib.error
import urllib.request
from dataclasses import dataclass
from pathlib import Path
from typing import Any


@dataclass(frozen=True)
class PipClientResult:
    enabled: bool
    available: bool
    fallback_used: bool
    data: dict[str, Any] | None = None
    error_code: str | None = None
    error_message: str | None = None
    latency_ms: int | None = None


def _env_bool(name: str, default: bool = False) -> bool:
    raw = os.getenv(name)
    if raw is None:
        return default
    return raw.strip().lower() in {"1", "true", "yes", "on"}


class PipClient:
    def __init__(self) -> None:
        self.integration_enabled = _env_bool("PIP_INTEGRATION_ENABLED", False)
        self.required_for_recommendations = _env_bool("PIP_REQUIRED_FOR_RECOMMENDATIONS", False)
        self.failsafe_mode = _env_bool("PIP_FAILSAFE_MODE", True)
        self.use_mock = _env_bool("PIP_USE_MOCK", False)
        self.base_url = os.getenv("PIP_BASE_URL", "").rstrip("/")
        self.api_key = os.getenv("PIP_API_KEY", "")
        try:
            self.timeout_seconds = max(1.0, float(os.getenv("PIP_TIMEOUT_SECONDS", "5")))
        except ValueError:
            self.timeout_seconds = 5.0

    def get_fixture_probability(self, fixture_id: int, market: str = "1X2") -> PipClientResult:
        if not self.integration_enabled:
            return PipClientResult(False, False, True, error_code="PIP_DISABLED", error_message="PIP integration is disabled by configuration.")
        if self.use_mock:
            return self._load_mock_response(fixture_id=fixture_id, market=market)
        if not self.base_url:
            return self._fail("PIP_BASE_URL_MISSING", "PIP_BASE_URL is not configured.")

        endpoint = f"{self.base_url}/api/v1/probability/football/fixture/{int(fixture_id)}?market={market}"
        headers = {"Accept": "application/json"}
        if self.api_key:
            headers["X-PIP-API-Key"] = self.api_key

        start = time.perf_counter()
        try:
            request = urllib.request.Request(endpoint, headers=headers, method="GET")
            with urllib.request.urlopen(request, timeout=self.timeout_seconds) as response:
                latency_ms = int((time.perf_counter() - start) * 1000)
                if response.status != 200:
                    return self._fail("PIP_HTTP_ERROR", f"PIP returned HTTP {response.status}", latency_ms)
                payload = json.loads(response.read().decode("utf-8"))
                if payload.get("execution_allowed") is not False:
                    return self._fail("PIP_CONTRACT_EXECUTION_FLAG", "PIP response violated execution_allowed=false contract.", latency_ms)
                return PipClientResult(True, True, False, data=payload, latency_ms=latency_ms)
        except urllib.error.HTTPError as exc:
            return self._fail("PIP_HTTP_ERROR", f"PIP HTTP error: {exc.code}")
        except urllib.error.URLError as exc:
            return self._fail("PIP_UNREACHABLE", f"PIP unreachable: {exc.reason}")
        except TimeoutError:
            return self._fail("PIP_TIMEOUT", "PIP request timed out.")
        except Exception as exc:
            return self._fail("PIP_CLIENT_EXCEPTION", str(exc))

    def _load_mock_response(self, fixture_id: int, market: str) -> PipClientResult:
        start = time.perf_counter()
        root = Path(__file__).resolve().parents[2]
        mock_path = root / "examples" / "mock_pip_probability_response.json"
        try:
            payload = json.loads(mock_path.read_text(encoding="utf-8"))
            payload["fixture_id"] = int(fixture_id)
            payload["market"] = market
            latency_ms = int((time.perf_counter() - start) * 1000)
            return PipClientResult(True, True, False, payload, latency_ms=latency_ms)
        except Exception as exc:
            return self._fail("PIP_MOCK_LOAD_FAILED", str(exc))

    def _fail(self, code: str, message: str, latency_ms: int | None = None) -> PipClientResult:
        if not self.failsafe_mode and self.required_for_recommendations:
            raise RuntimeError(f"{code}: {message}")
        return PipClientResult(True, False, True, error_code=code, error_message=message, latency_ms=latency_ms)
